graviton/invariant.py

```diff
@@ -10,14 +11,47 @@
 )
 
 
+class PredicateKind(Enum):
+    Constant = "constant"
+    CaseMap = "case map"
+    ExactMatch = "exact match"
+    RangeMatch = "range match"
+    IdenticalPair = "identical"
+
+
 INVARIANT_TYPE = Union[
     PyTypes,
     Dict[Sequence[PyTypes], PyTypes],
     Callable[[PyTypes], PyTypes],
     Callable[[PyTypes], bool],
+    PredicateKind,  # currently represents the equality of two dry-run executable functions
 ]
 
 
+def get_kind(predicate: INVARIANT_TYPE) -> PredicateKind:
+    if isinstance(predicate, PredicateKind):
+        return predicate
+
+    if isinstance(predicate, dict):
+        return PredicateKind.CaseMap
+
+    if not callable(predicate):
+        return PredicateKind.Constant
+
+    try:
+        sig = signature(predicate)
+    except Exception as e:
+        raise Exception(f"callable predicate {predicate} must have a signature") from e
+
+    N = len(sig.parameters)
+    assert N in (1, 2), f"predicate has the wrong number of paramters {N}"
+
+    if N == 2:
+        return PredicateKind.RangeMatch
+
+    return PredicateKind.ExactMatch
```

Comment:
Another dumb question: the callable object seems to be having only 1 or 2 argument, could you pin point me any examples of `INVARIANT_TYPE` falling into the 2 categories?

Reply:
1. [cost lambda](https://github.com/algorand/graviton/blob/2c562cc601b85392a57d8bc4c4b6b7c32d224432/tests/integration/blackbox_test.py#L321) with a single confusingly named `args` argument
2. [final scratch lambda](https://github.com/algorand/graviton/blob/2c562cc601b85392a57d8bc4c4b6b7c32d224432/tests/integration/blackbox_test.py#L329) with first parameter `args` and second parameter `actual`